File: uni_api/providers/overrides.py
from __future__ import annotations

import copy

from core.utils import get_model_dict, safe_get


def _path_parts(path: object) -> list[str]:
    if not isinstance(path, str):
        return []
    return [part for part in (part.strip() for part in path.split(".")) if part]


def _get_path(value: object, path: str) -> tuple[object, bool]:
    current = value
    for part in _path_parts(path):
        if not isinstance(current, dict) or part not in current:
            return None, False
        current = current[part]
    return current, True


def _delete_path(value: dict, path: str) -> None:
    parts = _path_parts(path)
    if not parts:
        return
    current = value
    for part in parts[:-1]:
        child = current.get(part)
        if not isinstance(child, dict):
            return
        current = child
    current.pop(parts[-1], None)


def _set_path(value: dict, path: str, replacement: object) -> None:
    parts = _path_parts(path)
    if not parts:
        return
    current = value
    for part in parts[:-1]:
        child = current.get(part)
        if not isinstance(child, dict):
            return
        current = child
    current[parts[-1]] = replacement


def _value_matches_expected(actual: object, expected: object) -> bool:
    return actual == expected


def _matches_condition(value: object, condition: object) -> bool:
    if isinstance(condition, str):
        return value == condition
    if isinstance(condition, dict):
        for key, expected in condition.items():
            actual, exists = _get_path(value, str(key))
            if not exists or not _value_matches_expected(actual, expected):
                return False
        return True
    return False


def _matches_any_condition(value: object, conditions: object) -> bool:
    if isinstance(conditions, (list, tuple, set)):
        return any(_matches_condition(value, condition) for condition in conditions)
    return _matches_condition(value, conditions)


def _apply_structured_removal(payload: dict, rule: dict) -> None:
    path = rule.get("path")
    if not isinstance(path, str) or not path.strip():
        return

    target, exists = _get_path(payload, path)
    if not exists:
        return

    has_where = "where" in rule
    has_where_any = "where_any" in rule

    if not has_where and not has_where_any:
        _delete_path(payload, path)
        return

    def should_remove(value: object) -> bool:
        return (
            (has_where and _matches_condition(value, rule.get("where")))
            or (has_where_any and _matches_any_condition(value, rule.get("where_any")))
        )

    if isinstance(target, list):
        filtered = [item for item in target if not should_remove(item)]
        if len(filtered) == len(target):
            return
        if filtered or not rule.get("drop_empty", False):
            _set_path(payload, path, filtered)
        else:
            _delete_path(payload, path)
        return

    if should_remove(target):
        _delete_path(payload, path)


def _merge_post_body_override_dict(target: dict, override: dict, *, removal_key: str = "__remove__") -> None:
    for key, value in override.items():
        if key == removal_key:
            continue
        current = target.get(key)
        if isinstance(current, dict) and isinstance(value, dict):
            _merge_post_body_override_dict(current, value, removal_key=removal_key)
        else:
            target[key] = copy.deepcopy(value)


def apply_post_body_parameter_overrides(
    payload: dict,
    provider: dict,
    request_model: str,
    *,
    skip_keys: set[str] | None = None,
) -> dict:
    removal_key = "__remove__"

    def normalize_removals(value) -> list[str | dict]:
        if isinstance(value, str):
            cleaned = value.strip()
            return [cleaned] if cleaned else []
        if isinstance(value, (list, tuple, set)):
            removals: list[str | dict] = []
            for item in value:
                if isinstance(item, str):
                    cleaned = item.strip()
                    if cleaned:
                        removals.append(cleaned)
                elif isinstance(item, dict):
                    removals.append(item)
            return removals
        if isinstance(value, dict):
            return [value]
        return []

    overrides = safe_get(provider, "preferences", "post_body_parameter_overrides", default={}) or {}
    if not isinstance(overrides, dict):
        return payload

    model_dict = provider.get("_model_dict_cache")
    if not isinstance(model_dict, dict):
        model_dict = get_model_dict(provider)

    skipped = skip_keys or set()

    def apply_override_values(override: dict, *, skip_model_override_keys: bool = False) -> None:
        for key, value in override.items():
            if (
                key in skipped
                or key == removal_key
                or (skip_model_override_keys and key in model_dict)
            ):
                continue
            current = payload.get(key)
            if isinstance(current, dict) and isinstance(value, dict):
                _merge_post_body_override_dict(current, value, removal_key=removal_key)
            else:
                payload[key] = copy.deepcopy(value)

    def apply_removals(override: dict) -> None:
        for removal in normalize_removals(override.get(removal_key)):
            if isinstance(removal, str):
                key = removal
            elif isinstance(removal, dict):
                key = removal.get("path")
            else:
                continue
            if key in skipped:
                continue
            if isinstance(removal, dict):
                _apply_structured_removal(payload, removal)
            else:
                payload.pop(key, None)

    apply_override_values(overrides, skip_model_override_keys=True)
    apply_removals(overrides)

    model_overrides = overrides.get(request_model)
    if isinstance(model_overrides, dict):
        apply_override_values(model_overrides)
        apply_removals(model_overrides)

    return payload

File: src/providers/overrides.rs
use crate::config::snapshot::Provider;
use serde_json::{Map, Value};

pub(crate) fn apply_overrides(
    root: &mut Map<String, Value>,
    provider: &Provider,
    request_model: &str,
) {
    if let Some(overrides) = provider
        .preferences
        .get("post_body_parameter_overrides")
        .and_then(Value::as_object)
    {
        apply_override_section(root, overrides, provider, true);
        if let Some(model) = overrides.get(request_model).and_then(Value::as_object) {
            apply_override_section(root, model, provider, false);
        }
    }
    // Codex wire requirements also apply when no overrides are configured,
    // and cannot be undone by provider-wide or model-specific overrides.
    // Endpoint-specific sanitizers still run afterwards (e.g. compact drops store).
    if provider.engine.trim().eq_ignore_ascii_case("codex") {
        root.insert("store".into(), Value::Bool(false));
        root.remove("response_format");
        root.remove("temperature");
    }
}

pub(crate) fn apply_override_section(
    root: &mut Map<String, Value>,
    section: &Map<String, Value>,
    provider: &Provider,
    skip_model_keys: bool,
) {
    for (key, value) in section {
        if key == "__remove__"
            || matches!(key.as_str(), "service_tier" | "translation_options")
            || (skip_model_keys && provider.models.contains_key(key))
        {
            continue;
        }
        merge_value(root.entry(key.clone()).or_insert(Value::Null), value);
    }
    if let Some(removals) = section.get("__remove__") {
        apply_removals(root, removals);
    }
}

pub(crate) fn merge_value(target: &mut Value, replacement: &Value) {
    match (target, replacement) {
        (Value::Object(target), Value::Object(replacement)) => {
            for (key, value) in replacement {
                if key == "__remove__" {
                    continue;
                }
                merge_value(target.entry(key.clone()).or_insert(Value::Null), value);
            }
        }
        (target, replacement) => *target = replacement.clone(),
    }
}

pub(crate) fn apply_removals(root: &mut Map<String, Value>, removals: &Value) {
    let items = match removals {
        Value::Array(items) => items.clone(),
        value => vec![value.clone()],
    };
    for item in items {
        if let Some(path) = item.as_str() {
            if !matches!(path, "service_tier" | "translation_options") {
                delete_path(root, path);
            }
        } else if let Some(rule) = item.as_object() {
            if !matches!(
                rule.get("path").and_then(Value::as_str),
                Some("service_tier" | "translation_options")
            ) {
                apply_structured_removal(root, rule);
            }
        }
    }
}

pub(crate) fn apply_structured_removal(root: &mut Map<String, Value>, rule: &Map<String, Value>) {
    let Some(path) = rule.get("path").and_then(Value::as_str) else {
        return;
    };
    if !rule.contains_key("where") && !rule.contains_key("where_any") {
        delete_path(root, path);
        return;
    }
    let Some(target) = get_path_mut(root, path) else {
        return;
    };
    let should_remove = |value: &Value| {
        rule.get("where")
            .is_some_and(|condition| matches_condition(value, condition))
            || rule
                .get("where_any")
                .is_some_and(|condition| matches_any_condition(value, condition))
    };
    if let Some(items) = target.as_array_mut() {
        items.retain(|item| !should_remove(item));
        if items.is_empty() && rule.get("drop_empty").and_then(Value::as_bool) == Some(true) {
            delete_path(root, path);
        }
    } else if should_remove(target) {
        delete_path(root, path);
    }
}

pub(crate) fn matches_any_condition(value: &Value, condition: &Value) -> bool {
    condition
        .as_array()
        .map(|conditions| conditions.iter().any(|item| matches_condition(value, item)))
        .unwrap_or_else(|| matches_condition(value, condition))
}

pub(crate) fn matches_condition(value: &Value, condition: &Value) -> bool {
    if condition.is_string() {
        return value == condition;
    }
    let Some(condition) = condition.as_object() else {
        return false;
    };
    condition.iter().all(|(path, expected)| {
        get_value_path(value, path).is_some_and(|actual| actual == expected)
    })
}

pub(crate) fn delete_path(root: &mut Map<String, Value>, path: &str) {
    let parts = path
        .split('.')
        .map(str::trim)
        .filter(|part| !part.is_empty())
        .collect::<Vec<_>>();
    if parts.is_empty() {
        return;
    }
    let mut current = root;
    for part in &parts[..parts.len() - 1] {
        let Some(next) = current.get_mut(*part).and_then(Value::as_object_mut) else {
            return;
        };
        current = next;
    }
    current.remove(parts[parts.len() - 1]);
}

pub(crate) fn get_path_mut<'a>(
    root: &'a mut Map<String, Value>,
    path: &str,
) -> Option<&'a mut Value> {
    let mut parts = path.split('.').filter(|part| !part.trim().is_empty());
    let first = parts.next()?;
    let mut current = root.get_mut(first)?;
    for part in parts {
        current = current.as_object_mut()?.get_mut(part)?;
    }
    Some(current)
}

pub(crate) fn get_value_path<'a>(value: &'a Value, path: &str) -> Option<&'a Value> {
    let mut current = value;
    for part in path.split('.').filter(|part| !part.trim().is_empty()) {
        current = current.as_object()?.get(part)?;
    }
    Some(current)
}
